Feature/known gaps

abdulrabbani00 opened the pull request:
# Overview
This PR contains does the following:

* Contains a function for updating `known_gaps` to the DB - `FindAndUpdateGaps`.
  *  This function has a unit-test: `TestKnownGapsUpsert`
* Triggering the function `FindAndUpdateGaps` throughout the codebase to check for gaps.

# Future Cases

1. `processingKey` - A field that can be used to differentiate the block calculation logic between geth nodes. We should be able to calculate the `expectedDiferrence` from the `processingKey` (perhaps through a mapping).
2. `expectedDifference` - Indicates what the difference between blocks should be.

Commit:
Function To Add entry to `known_gaps`

This commit contains all the code needed to update the `known_gaps` table. It uses a generic function `PushKnownGaps` to insert gaps to the database.

The current test is very "basic." All it does it insert a dummy gap into the database. As long as their is no error it will return succesfully.

I have also removed the foundry subfolder. I am using the `related-repository` setup for testing.

Commit:
Delete geth-linux-amd64

Commit:
Remove the need to pass a transaction when upserting to known_gaps

Thanks Ian :D

Commit:
Improve Testing for Insert, add a query function

Commit:
Find Gaps and at them to `known_gaps`

This contains all the logic needed to add a gap to the `known_gaps` table.

We can now add this code to various parts of the application to calculate if a gap has occured.

Commit:
Handle All Gaps within Geth

Including an updated doc which keeps track of events in this PR.

Commit:
% Checkpoint - Added the feature to write to File if writing to DB errors out. NOT TESTED

Commit:
** Write to file if you can't write to DB

This commit contains many new features:

1. Write the insert statement to file if there is an error when trying to write to DB.
2. Test the write to DB and write to file features in unit tests. They work as expected.
3. Insert logic for calling the knownGaps function (untested).

This PR makes a major change to the way statediff is started. Unless it is started in `file` mode, it will create a `file` indexer. This indexer will be used for writing errors to disk.

i-norden commented on cmd/geth/config.go:
```diff
@@ -191,6 +192,15 @@ func makeFullNode(ctx *cli.Context) (*node.Node, ethapi.Backend) {
+					FilePath: ctx.GlobalString(utils.StateDiffFilePath.Name),
+				}
+			} else {
+				fileConfig = nil
```
Is already `nil`, no need to set here

i-norden commented on statediff/indexer/constructor.go:
```diff
@@ -31,19 +31,27 @@ import (
+// You can specify the specific
+func NewStateDiffIndexer(ctx context.Context, chainConfig *params.ChainConfig, nodeInfo node.Info, config interfaces.Config, specificIndexer shared.DBType) (interfaces.StateDiffIndexer, error) {
+	var indexerToCreate shared.DBType
+	if specificIndexer == "" {
```
The indexer we create is already specified by the type of config we pass, we don't need this extra param to specify. `specificIndexer` has to match the type of `config.Type()` or we will get an error in the below switch statement when we try to assert the config interface as a concrete type.

i-norden commented on statediff/indexer/database/file/writer.go:
```diff
@@ -155,6 +155,14 @@ const (
 
 	storageInsert = "INSERT INTO eth.storage_cids (header_id, state_path, storage_leaf_key, cid, storage_path, " +
 		"node_type, diff, mh_key) VALUES ('%s', '\\x%x', '%s', '%s', '\\x%x', %d, %t, '%s');\n"
+	// INSERT INTO eth.known_gaps (starting_block_number, ending_block_number, checked_out, processing_key)
```
Let's clean out these comments if they aren't needed

i-norden commented on statediff/indexer/database/file/writer.go:
```diff
@@ -253,3 +261,12 @@ func (sqw *SQLWriter) upsertStorageCID(storageCID models.StorageNodeModel) {
+func (sqw *SQLWriter) upsertKnownGaps(knownGaps models.KnownGapsModel) {
+	sqw.stmts <- []byte(fmt.Sprintf(knownGapsInsert, knownGaps.StartingBlockNumber, knownGaps.EndingBlockNumber, knownGaps.CheckedOut, knownGaps.ProcessingKey,
+		knownGaps.EndingBlockNumber, knownGaps.ProcessingKey, knownGaps.EndingBlockNumber))
+	//knownGapsInsert = "INSERT INTO eth.known_gaps (starting_block_number, ending_block_number, checked_out, processing_key) " +
```
ditto

i-norden commented on statediff/indexer/database/file/writer.go:
```diff
@@ -253,3 +261,12 @@ func (sqw *SQLWriter) upsertStorageCID(storageCID models.StorageNodeModel) {
 		storageCID.Path, storageCID.NodeType, true, storageCID.MhKey))
 }
+
+func (sqw *SQLWriter) upsertKnownGaps(knownGaps models.KnownGapsModel) {
```
I'm not sure it makes sense to write out the inserts to the known_gaps table to the same file we are writing everything else to.

i-norden commented on statediff/indexer/database/sql/indexer.go:
```diff
@@ -553,3 +553,97 @@ func (sdi *StateDiffIndexer) PushCodeAndCodeHash(batch interfaces.Batch, codeAnd
+// REmove the return value
+// Write to file if err in writing to DB
+func (sdi *StateDiffIndexer) FindAndUpdateGaps(latestBlockOnChain *big.Int, expectedDifference *big.Int, processingKey int64, fileIndexer interfaces.StateDiffIndexer) error {
+	dbQueryString := "SELECT MAX(block_number) FROM eth.header_cids"
```
Let's move this to a private global constant

i-norden commented on statediff/service.go:
```diff
@@ -289,6 +348,75 @@ func (sds *Service) writeGenesisStateDiff(currBlock *types.Block, workerId uint)
+	}
+}
+
 func (sds *Service) writeLoopWorker(params workerParams) {
```
If we have multiple workers there are race conditions with the access to `sds.KnownGaps` in this method.

i-norden commented on statediff/service.go:
```diff
@@ -308,12 +436,48 @@ func (sds *Service) writeLoopWorker(params workerParams) {
+			// the actual lastProcessedBlock, add it to known gaps table.
+			if sds.KnownGaps.lastProcessedBlock != nil {
+				// Can't combine the two if statements because you can't compare a nil object...
+				if previousExpectedBlock.Cmp(sds.KnownGaps.lastProcessedBlock) != 0 {
```
And in the case of multiple workers the shared access to `sds.KnownGaps` means that this comparison will `!= 0` sometimes when it should because another worker updated `lastProcessedBlock`.

i-norden commented on statediff/service.go:
```diff
@@ -308,12 +436,48 @@ func (sds *Service) writeLoopWorker(params workerParams) {
+			}
+			sds.KnownGaps.lastProcessedBlock = currentBlock.Number()
+
 			// TODO: how to handle with concurrent workers
```
Ahh yeah you are aware, this should be considered at this stage because it will require a significant refactor. I'm starting to think this goal of trying to identify gaps that didn't occur due to some error is too complex given the need to support concurrent (out-of-order) processing and I can't really think of any cases where this would occur.

i-norden commented on cmd/geth/config.go:
```diff
@@ -191,6 +192,15 @@ func makeFullNode(ctx *cli.Context) (*node.Node, ethapi.Backend) {
+
+			if dbType != shared.FILE {
+				fileConfig = file.Config{
+					FilePath: ctx.GlobalString(utils.StateDiffFilePath.Name),
```
I think we should have a separate flag for the .sql file for the inserts to the `known_gaps` table, as when we are in file writing mode we should write those inserts to a different file than where all the statediff inserts are going. Thoughts?

i-norden replied:
This kind of relates to https://github.com/vulcanize/go-ethereum/pull/217#discussion_r836857914 I think the gap r/w process should be decoupled from the indexer.

i-norden commented on statediff/indexer/database/sql/indexer.go:
```diff
@@ -553,3 +553,97 @@ func (sdi *StateDiffIndexer) PushCodeAndCodeHash(batch interfaces.Batch, codeAnd
 }
+
+// Update the known gaps table with the gap information.
+func (sdi *StateDiffIndexer) PushKnownGaps(startingBlockNumber *big.Int, endingBlockNumber *big.Int, checkedOut bool, processingKey int64, fileIndexer interfaces.StateDiffIndexer) error {
```
Hmm not sure how I feel about passing an indexer interface to this indexer method, or more generally adding these `known_gap` table querying/writing methods to the indexer interface. These gap related methods could be encapsulated in a separate interface/type e.g. `GapWriter`.

Prior to these changes the indexer only writes and all the writes are batched in a single atomic Tx, whereas here we are introducing read interfaces and all our r/w access to the gaps table is outside of the batch Txs. So I think this can (and since it can, it should) be decoupled from the already complicated indexer.

If instead we have some `GapWriter` type, we can inititalize this type with a `sql.Database` and an `os.File` e.g.
```go
type GapWriter interface {
    FindAndUpdateGaps(latestBlockOnChain *big.Int, expectedDifference *big.Int, processingKey int64) error
    PushKnownGaps(startingBlockNumber *big.Int, endingBlockNumber *big.Int, checkedOut bool, processingKey int64) error
}
func NewGapWriter(otherParams Params, db sql.Database, file *os.File) GapWriter {...}
```

And it can handle routing to db vs file internally without needing to pass an entire indexer type on each method call.

i-norden replied:
This would also avoid having to support/maintain the gap writing process/methods across all three indexers (sql, file, dump). Instead, there could be a single `GapWriter` in its own file (I think it could exist in the root `statediff` pkg?) which would be embedded and called directly in the statediffing service.

abdulrabbani00 replied:
I like this, but I am not sure an interface is needed. Also, the `newGapWriter` function is tricky because I want to be able to create the `gapWrite` structure with select parameters. But I agree with the rest entirely.

i-norden commented on statediff/indexer/database/sql/indexer.go:
```diff
@@ -553,3 +553,97 @@ func (sdi *StateDiffIndexer) PushCodeAndCodeHash(batch interfaces.Batch, codeAnd
+	log.Info("Writing known gaps to the DB")
+	if err := sdi.dbWriter.upsertKnownGaps(knownGap); err != nil {
+		log.Warn("Error writing knownGaps to DB, writing them to file instead")
+		fileIndexer.PushKnownGaps(startingBlockNumber, endingBlockNumber, checkedOut, processingKey, nil)
```
`fileIndexer.PushKnownGaps()` adds the gap statement to the `file.SQLWriter.stmts` cache, but I don't see where that cache is being flushed (`file.SQLWriter.Flush()`). And this is another indication that we shouldn't be putting these gap r/w methods on the indexer interface and passing file indexers around. If all we are doing is writing out these relatively infrequent and small gap statements we don't need the additional overhead of the file indexer and the internal caching that its `file.SQLWriter` performs + all the additional complexity that adds.

i-norden reviewed (commented):
Looking good! Added some comments, the main issue I see is with respect to the race conditions we will run into when we have multiple writeLoop workers (how we run it in prod) which you've already noted. But I also very strongly think we should decouple these gap r/w methods from the Indexer interface.

abdulrabbani00 commented:
> Looking good! Added some comments, the main issue I see is with respect to the race conditions we will run into when we have multiple writeLoop workers (how we run it in prod) which you've already noted. But I also very strongly think we should decouple these gap r/w methods from the Indexer interface.

Thank you for the thorough review, a few notes:

* I strongly agree with decoupling the known gaps with the indexer. Honestly, I have no idea why I tried to jam known gaps in there so strongly, it gave me nothing but trouble (and pain).  I will work on refactoring.
* The race condition for multiple workers is certainly worrisome. I was considering the best way to proceed considering the write to `lastProcessedBlock`.  There is no simple fix in my mind just yet. 
  * Let's consider the only case in my mind where this would actually be needed - 
    * We processed blocks 10, 11, 12. 
    * For some reason, `geth` did not receive blocks 13, 14, 15.
    * We receive blocks 16, 17 18,.... and processed them
  * In this scenario, we would want to add blocks 13-15 to the known gaps table. The question I have is:
    *  Is the above-mentioned scenario even possible?
    *  Does `geth` have any build-in mechanism that would realize we jumped from block 12 to 16, and request block 13-15? Because if it does, then we do not need to handle this use case.
  * If `geth` does allow us to jump from blocks and have these gaps then we might need to consider an implementation.
  * It's also important to note that if we have any errors when processing 13-15, those blocks are added to known gaps.
  * If `geth` simply errors out and panics when we see we went from block 12 to 16, that's okay too, because when `geth` restarts it will see the last processed block was block 12, and we are currently on block 16+n, it will capture 13-16+n in the known gaps table.
* Depending on the above, testing will vary. We should talk about the possibility of undetected gaps within geth.

Commit:
Refactor: Decouple knownGaps and Indexer

This commit decouples knownGaps and Indexer. All knownGaps logic is in its own file. This makes testing and maintainability easier.

We have also removed all efforts to check the `lastProcessedblock` - This is because we won't ever run into this issue (hyptothetically), because geth won't let it happen.

Commit:
Add Prometheus metrics and write SQL file if it has data.

Commit:
Code Clean Up

i-norden commented on statediff/docs/database.md:
```diff
@@ -0,0 +1,21 @@
+
+This document will go through some notes on the database component of the statediff service.
+
+# Components
```
Nice! I think I'll expand on this section in a later PR.

i-norden replied:
Nice! I think I'll expand on this section in a later PR.

i-norden commented on statediff/indexer/database/sql/indexer_shared_test.go:
```diff
@@ -22,6 +23,7 @@ var (
 	db        sql.Database
 	err       error
 	ind       interfaces.StateDiffIndexer
+	chainConf = params.MainnetChainConfig
```
I think this is unused

i-norden replied:
I think this is unused

i-norden commented on statediff/indexer/database/sql/postgres/database.go:
```diff
@@ -36,6 +36,7 @@ type DB struct {
 }
 
 // InsertHeaderStm satisfies the sql.Statements interface
+// Stm == Statement
```
Thanks, at some point we should rename all these methods to use 'Stmt' instead of 'Stm'

i-norden replied:
Thanks, at some point we should rename all these methods to use 'Stmt' instead of 'Stm'

i-norden commented on statediff/indexer/models/models.go:
```diff
@@ -147,3 +147,11 @@ type LogsModel struct {
+	StartingBlockNumber string `db:"starting_block_number"`
+	EndingBlockNumber   string `db:"ending_block_number"`
+	CheckedOut          bool   `db:"checked_out"`
+	ProcessingKey       int64  `db:"processing_key"`
```
If we decide to do the `mod_block_number` + `offset` we'll need to reflect that here

i-norden replied:
If we decide to do the `mod_block_number` + `offset` we'll need to reflect that here

i-norden commented on statediff/known_gaps.go:
```diff
@@ -0,0 +1,278 @@
+}
+
+// Unused
+func NewKnownGapsState(checkForGaps bool, processingKey int64, expectedDifference *big.Int,
```
Keep the "unused" notification but should also add a godoc comment here

i-norden replied:
Keep the "unused" notification but should also add a godoc comment here

i-norden commented on statediff/known_gaps.go:
```diff
@@ -0,0 +1,278 @@
+
+func (kg *KnownGapsState) writeSqlFileStmtToDb() error {
+	log.Info("Writing the local SQL file for KnownGaps to the DB")
+	file, ioErr := ioutil.ReadFile(kg.writeFilePath)
```
I prefer to keep all errors named 'err' unless there is some variable shadowing issue

i-norden replied:
I prefer to keep all errors named 'err' unless there is some variable shadowing issue

i-norden reviewed (commented):
LGTM just some nitpicky comments, and if we decide to use `mod_block_number` + `offset` columns in the DB we'll need to update that here.

There is also some linting issue which I think is left over from my previous work, if you can fix that so that the linting CI test passes that'd be great.

Commit:
PR Updates as per Ian's request

Commit:
Update docker compose

Commit:
Update known_gaps table name

Commit:
Add Workflow dispatch for docker builds

i-norden reviewed (approved):
:shipit:
